### ISSUE TITLE
Compatibility with streaming APIs.  

### ISSUE THREAD

**Issue by garronej** [garronej/tss-react#125]
Hello,  

I'm very sorry to bother you but I have this burning question I wanted to ask you for weeks.  

Is there a world where `tss-react` is compatible with the streaming APIs?  

This module is a wrapper around `@emotion/react` that provide a hook-based API.  [See your quote](https://github.com/mui/material-ui/issues/24109#issuecomment-780855014).  

As of today, the only way to make tss-react work with SSR is to implement [the advanced approach](https://emotion.sh/docs/ssr#advanced-approach) and this approach doesn’t work with the steaming APIs.  

<img width="865" alt="image" src="https://user-images.githubusercontent.com/6702424/198372273-076dc65b-a100-4476-9941-43e5a6c9b4fa.png">

I don't care how hard it is to set up, I'm willing to provide and maintain a specific SSR utility on a per framework basis [just like I currently do for Next.js](https://docs.tss-react.dev/ssr/next.js).  

But is it or will it ever be feasible to support streaming?  

Thank you very much for taking a look at this.

**Comment by Andarist** [comment]
Yes, it should be possible to support streaming - if you only wrap your app with some special provider etc. Emotion itself is not 100% compatible with streaming right now, it's quite nuanced. I'm - sort of - waiting for better APIs (Resources) that are supposed to make it easier to implement this stuff in the future. 

As right now, to make streaming possible we'd have to render smth like that on the server (or inject this into the stream):
```
       <style
          // construct the style tag from the styles inserted into the cache since the lust chunk was written to the stream
        />
        <script>
          var cs = document.currentScript
          document.head.appendChild(cs.previousSibling)
          cs.remove()
        </script>
```

On top of that, we'd have to figure out how to rehydrate this properly on the client side (totally doable, I just didn't exactly think deeply about how to do this best).

With Next.js the new version also comes with a custom hook, `useServerInsertedHTML`, that could potentially be utilized to make this easier for us. However, from what I understand, this doesn't actually support streaming all that well because it only allows us to inject into the head **once**. I contacted the Next.js team this morning with a question about this. So I hope to learn more about this soon-ish.

Either way - this potentially could only make the integration with Next.js smoother and we need to support a wider ecosystem. If you'd like to collaborate on making a stream transformer for Emotion that could be used with any framework then I'm totally open to it. I have some idea on how we could make this work - I just couldn't find enough energy to do it on my own in the past months.

**Comment by garronej** [comment]
Thank you for the quick and detailed answer @Andarist,  

This is very encouraging, I'm very glad that `tss-react` is not dead tech.  

> I'm waiting for better APIs (Resources) that are supposed to make it easier to implement this stuff in the future.  

I'm not in a hurry at all, no user complained yet and I have no usecase for it.  
I just wanted to know if we'll eventually be able, one way or another to support it or if it was a dead end.  

> I just couldn't find enough energy to do it on my own in the past months.  

I'd love to but I don't think I'll have the bandwidth in the coming weeks.  
That said, please, if you open an issue about it, tag me, I'll see what I can do!  
I think waiting a little bit for the dust to settle around this is a good call.  

Thanks again @Andarist,
